sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/MultiServiceManager.java

```diff
@@ -0,0 +1,282 @@
+            }
+
+            if (isRegistered) {
+                // We are already registered. Manually call registered() against this client so that it can initialize.
+                service.registered(false);
+            }
+            return this;
+        } finally {
+            rwlock.unlock();
+        }
+    }
+
+    /**
+     * Triggers an uninstall for a service, removing it from the list of services when it has finished. Does nothing if
+     * the service is already uninstalling or doesn't exist. If the scheduler process is restarted, the service must be
+     * added again via {@link #putService(AbstractScheduler)}, at which point it will automatically resume uninstalling.
+     *
+     * @param serviceName the name of the service to be uninstalled
+     */
+    public void uninstallService(String serviceName) {
+        uninstallServices(Collections.singleton(serviceName));
+    }
+
+    /**
+     * Returns a service matching the provided {@code TaskStatus}, or an empty {@link Optional} if no match was found.
+     */
+    public Optional<AbstractScheduler> getMatchingService(Protos.TaskStatus status) {
+        Optional<String> sanitizedServiceName;
+        try {
+            sanitizedServiceName = CommonIdUtils.toSanitizedServiceName(status.getTaskId());
+        } catch (TaskException e) {
+            sanitizedServiceName = Optional.empty();
+        }
+        if (!sanitizedServiceName.isPresent()) {
+            // Bad task id.
+            LOGGER.error("Received task status with malformed id '{}', unable to route to service: {}",
+                    status.getTaskId().getValue(), TextFormat.shortDebugString(status));
+            return Optional.empty();
+        }
+
+        return getServiceSanitized(sanitizedServiceName.get());
+    }
+
+    /**
+     * UNINSTALL FLOW:
+     * 1. uninstallService("foo") is called. This converts the DefaultScheduler for the service to an UninstallScheduler
+     * 2. UninstallScheduler internally flags its StateStore with an uninstall bit if one is not already present.
+     * 3. The UninstallScheduler proceeds to clean up the service.
+     * 4. In the event of a scheduler process restart during cleanup:
+     *   a. Upstream builds a new foo using SchedulerBuilder, which internally finds the uninstall bit and returns a
+     *      new UninstallScheduler
+     *   b. putService(foo) is called with the UninstallScheduler
+     *   c. The UninstallScheduler resumes cleanup from where it left off...
+     * 5. Sometime after the UninstallScheduler finishes cleanup, it returns FINISHED in response to offers.
+     * 6. We remove the service and invoke uninstallCallback.uninstalled(), telling upstream that it's gone. If upstream
+     *    invokes putService(foo) again at this point, the service will be relaunched from scratch because the uninstall
+     *    bit in ZK will have been cleared.
+     */
+    public void uninstallServices(Collection<String> finishedServiceNames) {
```

Comment:
Because uninstall is going to be happening so frequently now, we should probably put a timeout on it.  If we're waiting for resource `abc-123` for more than 10 minutes or something, consider it gone and let uninstall proceed.  We don't want to have to handle many or very old uninstalls which are stuck.  Garbage collection will happen eventually anyway if resources from unexpected services ever show up again.

Reply:
INFINITY-3496